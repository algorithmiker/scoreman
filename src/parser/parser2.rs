use anyhow::{bail, Context};
use std::io::BufRead;

use crate::collect_parse_error;

use super::{comment_line, partline, Score, Section};

pub fn parse2<A: BufRead>(inp: A) -> anyhow::Result<Score> {
    let mut sections = vec![];
    // Todo eventually remove Part
    let mut part_buf = vec![];
    let mut part_begin = 0;
    for (line_idx, line) in inp.lines().enumerate() {
        let line_number = line_idx + 1;
        let line = line.with_context(|| format!("Cannot read line {line_number}"))?;
        if line.trim().is_empty() {
            if !part_buf.is_empty() {
                println!("[W]: Empty line inside Part at line {line_idx}, are you sure this is intended?");
            }
            continue;
        }

        match comment_line(&line) {
            Ok((rem, comment)) => {
                if !rem.is_empty() {
                    bail!("Invalid comment syntax at line {line_number} (got remaining `{rem}`)")
                }
                if !part_buf.is_empty() {
                    println!("[W]: Comment inside Part at line {line_idx}, are you sure this is intended?");
                }
                sections.push(Section::Comment(comment.to_string()));
            }
            Err(_) => match partline(&line) {
                Ok((rem, mut line)) => {
                    if !rem.is_empty() {
                        bail!("Invalid partline at line {line_number}, got remaining content: `{rem}`");
                    }
                    // Add measure metadata
                    for measure_idx in 0..line.staffs.len() {
                        line.staffs[measure_idx].parent_line = Some(line_idx);
                        line.staffs[measure_idx].index_on_parent_line = Some(measure_idx);
                    }
                    part_buf.push(line);
                    match part_buf.len() {
                        6 => {
                            // flush part buf
                            sections.push(Section::Part {
                                part: part_buf
                                    .clone() // TODO try to elide this clone
                                    .try_into()
                                    .expect("Unreachable: more than 6 elements in part_buf"),
                                begin_line_idx: part_begin,
                                end_line_idx: line_idx,
                            });
                            part_buf.clear();
                            part_begin = 0;
                        }
                        1 => part_begin = line_idx,
                        _ => (),
                    }
                }
                // TODO maybe pass the error up here?
                Err(x) => bail!(
                    "Error at line {line_number}, not a comment but also not a valid partline\n{}",
                    collect_parse_error(x)
                ),
            },
        }
    }

    Ok(Score(sections))
}

#[test]
fn test_parse2() {
    use std::io::BufReader;
    let i1 = r#"
e|---|
B|-3-|
G|6-6|
D|---|
A|---|
E|---|

// This is a comment 

e|---|
B|3-3|
G|-6-|
D|---|
A|---|
E|---|"#;
    insta::assert_debug_snapshot!(parse2(BufReader::new(i1.as_bytes())));
    let i2 = r#"
e|---|
B|-3-|
G|6-6|
D|---|
A|-o-|
E|---|

// This is a comment 

e|---|
B|3-3|
G|-6-|
D|---|
A|---|
E|---|"#;
    insta::assert_debug_snapshot!(parse2(BufReader::new(i2.as_bytes())));
}
