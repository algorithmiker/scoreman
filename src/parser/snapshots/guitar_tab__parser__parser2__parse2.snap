---
source: src/parser/parser2.rs
expression: parser.parse(i1.lines())
---
Ok(
    Parse2Result {
        diagnostics: [],
        sections: [
            Part {
                part: [
                    Partline {
                        string_name: 'e',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'B',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'G',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'D',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'A',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'E',
                        measures: 0..=0,
                    },
                ],
            },
            Comment(
                "// This is a comment",
            ),
            Part {
                part: [
                    Partline {
                        string_name: 'e',
                        measures: 1..=1,
                    },
                    Partline {
                        string_name: 'B',
                        measures: 1..=1,
                    },
                    Partline {
                        string_name: 'G',
                        measures: 1..=1,
                    },
                    Partline {
                        string_name: 'D',
                        measures: 1..=1,
                    },
                    Partline {
                        string_name: 'A',
                        measures: 1..=1,
                    },
                    Partline {
                        string_name: 'E',
                        measures: 1..=1,
                    },
                ],
            },
        ],
        string_names: [
            'e',
            'B',
            'G',
            'D',
            'A',
            'E',
        ],
        strings: [
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Fret(
                        3,
                    ),
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Fret(
                        3,
                    ),
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Fret(
                        3,
                    ),
                },
            ],
            [
                RawTick {
                    element: Fret(
                        6,
                    ),
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Fret(
                        6,
                    ),
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Fret(
                        6,
                    ),
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
            ],
        ],
        measures: [
            [
                Measure {
                    content: 0..=2,
                    parent_line: 1,
                    index_on_parent_line: 1,
                },
                Measure {
                    content: 3..=5,
                    parent_line: 10,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=2,
                    parent_line: 2,
                    index_on_parent_line: 1,
                },
                Measure {
                    content: 3..=5,
                    parent_line: 11,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=2,
                    parent_line: 3,
                    index_on_parent_line: 1,
                },
                Measure {
                    content: 3..=5,
                    parent_line: 12,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=2,
                    parent_line: 4,
                    index_on_parent_line: 1,
                },
                Measure {
                    content: 3..=5,
                    parent_line: 13,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=2,
                    parent_line: 5,
                    index_on_parent_line: 1,
                },
                Measure {
                    content: 3..=5,
                    parent_line: 14,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=2,
                    parent_line: 6,
                    index_on_parent_line: 1,
                },
                Measure {
                    content: 3..=5,
                    parent_line: 15,
                    index_on_parent_line: 1,
                },
            ],
        ],
        offsets: [
            [
                3,
                4,
                5,
                68,
                69,
                70,
            ],
            [
                10,
                11,
                12,
                75,
                76,
                77,
            ],
            [
                17,
                18,
                19,
                82,
                83,
                84,
            ],
            [
                24,
                25,
                26,
                89,
                90,
                91,
            ],
            [
                31,
                32,
                33,
                96,
                97,
                98,
            ],
            [
                38,
                39,
                40,
                103,
                104,
                105,
            ],
        ],
        bend_targets: {},
    },
)
